### PR TITLE
add eeprom stress testing script

### PR DESCRIPTION
**Pull request by knmcguire**
This PR is adding eeprom stresstesting. With first only a script in /utils/ with some graph functionality, but it would be good to have something that the nightly testing try every time.

### CODE REVIEW

File: utils/persistent_params_stresstesting.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
+# -*- coding: utf-8 -*-
+#
+# ,---------,       ____  _ __
+# |  ,-^-,  |      / __ )(_) /_______________ _____  ___
+# | (  O  ) |     / __  / / __/ ___/ ___/ __ `/_  / / _ \
+# | / ,--'  |    / /_/ / / /_/ /__/ /  / /_/ / / /_/  __/
+#    +------`   /_____/_/\__/\___/_/   \__,_/ /___/\___/
+#
+# Copyright (C) 2021 Bitcraze AB
```
Comment (jonasdn) [inline utils/persistent_params_stresstesting.py:9]: 2022

File: utils/persistent_params_stresstesting.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import time
+
+import cflib.crtp
+from cflib.crazyflie import Crazyflie
+from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
+from cflib.utils import uri_helper
+
+WITH_PLOTTING = False
+
+#uri = uri_helper.uri_from_env(default='usb://0')
```
Comment (jonasdn) [inline utils/persistent_params_stresstesting.py:38]: This line can be removed

File: tests/QA/test_param.py [hunk 3]
```diff
@@ -186,6 +186,80 @@ def state_cb(name, state):
...
+        with SyncCrazyflie(test_setup.device.link_uri) as scf:
+            # Get the names of all parameters that can be persisted
+            persistent_params = get_all_persistent_param_names(scf.cf)
+
+            assert persistent_params is not None
+
+            # Get the state of all persistent parameters
+            total_time=[]
+
+            over_maximum_persist_parameter_value= 300
```
Comment (jonasdn) [inline tests/QA/test_param.py:238]: some awkward whitespace here, could you run flake8 on the script?

File: tests/QA/test_param.py [hunk 4]
```diff
@@ -186,6 +186,80 @@ def state_cb(name, state):
...
+            wait_for_callback_event.wait()
+
+        def persist_parameter(cf, complete_param_name):
+            wait_for_callback_event = Event()
+
+            def is_stored_callback(complete_name, success):
+                wait_for_callback_event.set()
+
+            cf.param.persistent_store(complete_param_name, callback=is_stored_callback)
+            wait_for_callback_event.wait()
```
Comment (jonasdn) [inline tests/QA/test_param.py:206]: All these wait will wait forever if something goes wrong and block tests running in the Crazylab. Perhaps you could have a timeout and an assert?

```
assert wait_for_callback_event.wait(timeout=5)
```
Since `wait()` returns `False` on timeout.

File: tests/QA/test_param.py [hunk 5]
```diff
@@ -186,6 +186,80 @@ def state_cb(name, state):
...
+
+                    get_persistent_state(scf.cf, param_name)
+
+                    clear_persistent_parameter(scf.cf, param_name)
+                    
+                    get_persistent_state(scf.cf, param_name)
+                    
+                    total_time.append(time.time()-start_time)
+
+            assert total_time[1:].sum()/total_time[1:].len() > 0.5
```
Comment (jonasdn) [inline tests/QA/test_param.py:260]: This line begs for a comment, it is hard to parse

File: tests/QA/test_param.py [hunk 6]
```diff
@@ -186,6 +186,80 @@ def state_cb(name, state):
             with pytest.raises(AttributeError):
                 scf.cf.param.persistent_get_state(param, state_cb)
 
+    def test_param_persistent_stress(self, test_setup):
```
Comment (jonasdn) [inline tests/QA/test_param.py:189]: Please add a doc string explaining what this methods tests, and how.

**Review by jonasdn — CHANGES_REQUESTED** [review]

**Review by jonasdn — COMMENTED** [review]
So I think it is difficult, but maybe this script (`persistent_params_stresstesting.py)` should be in crazyflie-python-lib instead. If it is not used by anything else in testing. Since crazyflie-testing is infrastructure for tests.

**Comment by knmcguire** [comment]
hi! thanks for your comments! It is still in progress as you can see but it helps fixing it quicker.

 What do you actually mean  with  'it is not used by anything else in testing'? I'm not sure if the crazyflie-python-lib is good for it either as it is purely diagnostic tool only.

**Comment by knmcguire** [comment]
arg... I did the unthinkable and had to force push :smile: accidentally amended something to an commit I already pushed to remote. 

Anyway, this test passes now with the current firmware! Just now is the question of what to do with the diagnostics script...

**Comment by tobbeanton** [comment]
> arg... I did the unthinkable and had to force push smile accidentally amended something to an commit I already pushed to remote.
> 
> Anyway, this test passes now with the current firmware! Just now is the question of what to do with the diagnostics script...

I vote for keeping it in this repo. It is in a utils folder which puts it a bit outside the framework. Maybe put it in a tools folder to make it even more clear?

**Review by tobbeanton — CHANGES_REQUESTED** [review]
Put utils/persistent_params_stresstesting.py in tools/persistent_params_stresstesting.py to make it even more clear that it is a separate thing.

File: tests/QA/test_param.py [hunk 7]
```diff
@@ -186,8 +184,72 @@ def state_cb(name, state):
             with pytest.raises(AttributeError):
                 scf.cf.param.persistent_get_state(param, state_cb)
 
+    # Stresstest the eeprom by setting, clearing and
+    #  setting the persistent parameter. This will create
+    #  holes in the eeprom memory which needs to be defragmented
```
Comment (jonasdn) [inline tests/QA/test_param.py:189]: nit: some indent issues

File: tests/QA/test_param.py [hunk 8]
```diff
@@ -186,8 +184,72 @@ def state_cb(name, state):
...
+                    start_time = time.time()
+                    scf.cf.param.get_value(param_name)
+                    scf.cf.param.persistent_store(param_name, is_stored_callback)
+                    scf.cf.param.persistent_clear(param_name, is_stored_cleared)
+                    scf.cf.param.persistent_store(param_name, is_stored_callback)
+                    total_time.append(time.time() - start_time)
+
+            # Assert when the average time to get, set, get, clear, get,
+            # and set a persist parameter takes longer than 0.5 seconds,
+            average_time = sum(total_time[1:]) / len(total_time[1:])
```
Comment (jonasdn) [inline tests/QA/test_param.py:237]: Nicely structured code!

File: tests/QA/test_param.py [hunk 9]
```diff
@@ -186,8 +184,71 @@ def state_cb(name, state):
...
+            assert average_time < max_avg_sec_per_parameter
+
+            # Due to the swisscheese in memory due to this test
+            # we can add a test for how long the dfrac action takes
+            # to push all the memory block close to eachother.
+            # This should not take longer than 2 seconds.
+            assert max(total_time[1:]) < max_sec_defrac
+
```
Comment (jonasdn) [inline tests/QA/test_param.py:245]: Nice comments!

File: tests/QA/test_param.py [hunk 10]
```diff
@@ -186,8 +184,71 @@ def state_cb(name, state):
...
+        with SyncCrazyflie(test_setup.device.link_uri) as scf:
+            # Get the names of all parameters that can be persisted
+            persistent_params = get_all_persistent_param_names(scf.cf)
+            assert persistent_params is not None
+
+            total_time = []
+            over_maximum_persist_parameter_value = 300
+            len_persist_params = len(persistent_params)
+            loop_iteration = over_maximum_persist_parameter_value // len_persist_params
+
```
Comment (jonasdn) [inline tests/QA/test_param.py:223]: Nicely structured!

File: tests/QA/test_param.py [hunk 11]
```diff
@@ -16,28 +16,26 @@
 import logging
 import time
 import random
+from threading import Event
```
Comment (jonasdn) [inline tests/QA/test_param.py:19]: Where is this used in the test?

**Review by jonasdn — CHANGES_REQUESTED** [review]
nice work!

File: tools/persistent_params_stresstesting.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
+# -*- coding: utf-8 -*-
```
Comment (jonasdn) [inline tools/persistent_params_stresstesting.py:1]: So this starts a new thing, with manual test scripts being in `tools/` maybe this should be added to docs in some ways? What goes in tools and why?

File: tests/QA/test_param.py (hunk 11 above)
Comment (knmcguire) [reply tests/QA/test_param.py:19]: it is used now :)

File: tools/persistent_params_stresstesting.py (hunk 12 above)
Comment (knmcguire) [reply tools/persistent_params_stresstesting.py:1]: not sure about that. This is in fact a diagnostics tool so I'm not sure if this needs to be explained in the tools much if others are not using it. We shouldn't document everything, as we don't for the python libraries

Comment (knmcguire) [reply tools/persistent_params_stresstesting.py:1]: Perhaps better explanation in the script itself perhaps?

Comment (tobbeanton) [reply tools/persistent_params_stresstesting.py:1]: I think improving the explanation in the script itself is a good middle ground.
